src/addons/metaservice/src/main/java/ee/ria/xroad/proxy/serverproxy/RestMetadataServiceHandlerImpl.java

```diff
@@ -140,6 +151,89 @@ public void startHandling(HttpServletRequest servletRequest, ProxyMessage reques
         opMonitoringData.setAssignResponseOutTsToResponseInTs(true);
     }
 
+    private void handleListMethods(ProxyMessage requestProxyMessage) throws IOException {
+        restResponse.getHeaders().add(new BasicHeader(MimeUtils.HEADER_CONTENT_TYPE, MimeTypes.JSON));
+        MethodListType methodList = OBJECT_FACTORY.createMethodListType();
+        methodList.getService().addAll(ServerConf.getServicesByDescriptionType(
+                requestProxyMessage.getRest().getServiceId().getClientId(), DescriptionType.OPENAPI3));
+        MAPPER.writeValue(restResponseBody, methodList);
+    }
+
+    private void handleAllowedMethods(ProxyMessage requestProxyMessage) throws IOException {
+        restResponse.getHeaders().add(new BasicHeader(MimeUtils.HEADER_CONTENT_TYPE, MimeTypes.JSON));
+        MethodListType methodList = OBJECT_FACTORY.createMethodListType();
+        methodList.getService().addAll(ServerConf.getAllowedServicesByDescriptionType(
+                requestProxyMessage.getRest().getServiceId().getClientId(),
+                requestProxyMessage.getRest().getClientId(),
+                DescriptionType.OPENAPI3));
+        MAPPER.writeValue(restResponseBody, methodList);
+    }
+
+    private void handleGetOpenApi(HttpServletRequest servletRequest, ProxyMessage requestProxyMessage)
+            throws URISyntaxException, IOException {
+        // parse query string
+        String fullURL = getFullURL(servletRequest) + "?" + requestProxyMessage.getRest().getQuery();
+        log.trace("fullURL={}", fullURL);
+        List<NameValuePair> pairs = URLEncodedUtils.parse(new URI(fullURL), Charset.forName("UTF-8"));
```

Comment:
There is no need to create the full URI, one can parse the query string part only:
```suggestion
        List<NameValuePair> pairs = URLEncodedUtils.parse(requestProxyMessage.getRest().getQuery(), StandardCharsets.UTF_8);
```
Also, using servletRequest becomes unnecessary.

Reply:
Updated